File: javascript/math/566. Reshape the Matrix.js
/**
 * @param {number[][]} mat
 * @param {number} r
 * @param {number} c
 * @return {number[][]}
 */
var matrixReshape = function(mat, r, c) {
    if (mat.length * mat[0].length != r * c) return mat;

    const flat_mat = mat.flat();
    const res = [];
    
    for (let i=0; i<r; i++) {
        const row = [];
        for (let j=0; j<c; j++) row.push(flat_mat[c*i + j]);

        res.push(row);
    }
    return res;
};

File: python/math/566. Reshape the Matrix.py
'''
(own)
Time: O(n)
Space: O(n)
'''

class Solution:
    def matrixReshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        # check illegal
        if len(mat) * len(mat[0]) != r * c:
            return mat

        flat = [item for row in mat for item in row]
        res = []
        for i in range(r):
            row = []
            for j in range(c):
                row.append(flat[c*i+j])

            res.append(row)

        return res